src/main/java/com/pinterest/secor/uploader/HadoopS3UploadManager.java

```diff
@@ -60,12 +61,23 @@ public HadoopS3UploadManager(SecorConfig config) {
         LOG.info("uploading file {} to {}", localLogFilename, logFileName);
 
         final Future<?> f = executor.submit(new Runnable() {
+        	final int totalRetries = Integer.parseInt(mConfig.getUploaderRetries());
             @Override
             public void run() {
+            	int retriesLeft = totalRetries;
+            	boolean success = false;
+            	while (retriesLeft > 0 && !success) {
                 try {
                     FileUtil.moveToCloud(localLogFilename, logFileName);
+            			success = true;
                 } catch (IOException e) {
-                    throw new RuntimeException(e);
+            			success = false;
+            			retriesLeft--;
```

Comment:
Usually people add exponential backup delay between retries (e.g. 100ms, 200ms, 400ms ...) or adding delay with random intervals if you have a lot of workers concurrently running facing similar retries.